### ISSUE TITLE
JavaFX and MathML.

### ISSUE THREAD

**Issue by scientificware** [javafxports/openjdk-jfx#71]
When I discovered that JavaFX could manage MathML, I also discovered that it was buggy. Becauseof MathML is rendered with WebKit, I compared with other WebKit based web browsers. I also noticed that since 2016 that those web browser make good progress with MaltML display. First I thought that the differences were due to outdated WebKit version used by JavaFX but last releases, with an uptodate WebKit version, show that the problem is the same.

I contacted F. Wang from Igalia who worked to port or improve MathML support in FireFox and later in WebKit. He confirms that it was a bug in JavaFX not in WebKit witch displays MathML correctly.

As he suggested me, I wrote to the JavaFX mailing list and had a short discussion about this issue on OpenJDK.Java.net :

http://mail.openjdk.java.net/pipermail/openjfx-dev/2017-December/021112.html
https://bugs.openjdk.java.net/browse/JDK-8147476

So, as suggested by mainteners, I'm trying to find by myself what's wrong with MathML in javaFX HTMLEditor.

- As far as I could read in javaFX java code,
  - I understand that HTMLEditor and WebView only manage and display a cached page.
  - I know that JavaFX use WebKit to display a webpage.
- So I don't understand why, 
  - when I tested webkit with Safari on OSX and IOS last versions, in an Apple Store and with my Ipod (with an oudated IOS 9), the display is pretty good :
![ipod_ios9_screencapture](https://user-images.githubusercontent.com/19194678/39093477-8d57983c-4620-11e8-9878-f98f945149c4.PNG)
  -  And with JavaFX HTMLEditor the display is pretty bad (https://github.com/scientificware/javafxmathml) : 
![screen_javafx_mathml](https://user-images.githubusercontent.com/19194678/39093773-2aea4de8-4625-11e8-8cce-407416796205.png)

I tested all applications with Mozilla MathML Torture Test 
(https://developer.mozilla.org/fr/docs/Mozilla/MathML_Project/MathML_Torture_Test) 

Could someone help me ?
- To find in which code level is the problem.
- I don't think the problem is in the JavaFX java code, I read it several times.
- I tried to find what wrong in the native code but, I don't know where to start !
- So where I should start to work ?

**Comment by scientificware** [comment]
I think that the problem is related only with the vertical size of each element :
I notice that the size of the cursor is right outside the mathematics text but when we move it through the mathematics text, the cursor size is far much lower than text size.
All horizontal sizes seem good, by example a <mtext> followed by a <mrow> in a <math> element :
![screen_javafx_mathml_2](https://user-images.githubusercontent.com/19194678/39094553-91175ec4-4631-11e8-927f-f119ef90e0c3.png)

Which part of code deals with vertical size ?

**Comment by scientificware** [comment]
Thank you very much for your help, and all these informations !
Thanks given me an entry point to investigate :-)

**Comment by scientificware** [comment]
First working comments :
- In fact, what did I need to start was just the confirmation that the render logic for math is only the webkit logic.
- Thanks to Ararunprasad's comment, I'm a bit reassure.
- The most important information that he gave, is that JavaFX doesn't have is own render implementation for webkit
- However, I not agree about the conclusion for the pointed code. In fraction rendering the openfont math table provides only gap value beetween numerator and denominator and the line thickness which separate numerator and denominator.  I'm afraid the logic is good, because all these values are cumulative an positive, eventhough with a null gap between and denominator and a thickness to 0, both elements can't overlap :  I use a similar method in Swing to implement a fraction View. May be the tickness default method ( ruleThicknessFallback() ) returns a negative value (must be verified) : this method seems to use the font current size to calculate the thickness.
- Until, I verify some things, I maintain my first analysis : only one problem affects all math elements in the same time : somewhere in the code, logic looses font or children vertical size.

**Comment by scientificware** [comment]
public class JavaFXMathML extends Application {
       ^
1 error
Exception in Application start method
java.lang.reflect.InvocationTargetException
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:564)
        at javafx.graphics/com.sun.javafx.application.LauncherImpl.launchApplicationWithArgs(LauncherImpl.java:464)
        at javafx.graphics/com.sun.javafx.application.LauncherImpl.launchApplication(LauncherImpl.java:363)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:564)
        at java.base/sun.launcher.LauncherHelper$FXHelper.main(LauncherHelper.java:941)
Caused by: java.lang.RuntimeException: Exception in Application start method
        at javafx.graphics/com.sun.javafx.application.LauncherImpl.launchApplication1(LauncherImpl.java:900)
        at javafx.graphics/com.sun.javafx.application.LauncherImpl.lambda$launchApplication$2(LauncherImpl.java:195)
        at java.base/java.lang.Thread.run(Thread.java:844)
Caused by: java.lang.IllegalAccessError: class com.sun.javafx.scene.control.Logging (in module javafx.controls) cannot access class com.sun.javafx.logging.PlatformLogger (in module javafx.base) because module javafx.base does not export com.sun.javafx.logging to module javafx.controls
        at javafx.controls/com.sun.javafx.scene.control.Logging.getControlsLogger(Logging.java:47)
        at javafx.controls/javafx.scene.control.Control$2.invalidated(Control.java:316)
        at javafx.base/javafx.beans.property.ObjectPropertyBase.markInvalid(ObjectPropertyBase.java:112)
        at javafx.base/javafx.beans.property.ObjectPropertyBase.set(ObjectPropertyBase.java:147)
        at javafx.graphics/javafx.css.StyleableObjectProperty.set(StyleableObjectProperty.java:82)
        at javafx.controls/javafx.scene.control.Control$2.set(Control.java:250)
        at javafx.controls/javafx.scene.control.Control$2.set(Control.java:233)
        at javafx.controls/javafx.scene.control.Control.loadSkinClass(Control.java:757)
        at javafx.controls/javafx.scene.control.Control$5.invalidated(Control.java:680)
        at javafx.base/javafx.beans.property.StringPropertyBase.markInvalid(StringPropertyBase.java:110)
        at javafx.base/javafx.beans.property.StringPropertyBase.set(StringPropertyBase.java:145)
        at javafx.graphics/javafx.css.StyleableStringProperty.set(StyleableStringProperty.java:83)
        at javafx.controls/javafx.scene.control.Control$5.set(Control.java:672)
        at javafx.graphics/javafx.css.StyleableStringProperty.applyStyle(StyleableStringProperty.java:69)
        at javafx.graphics/javafx.css.StyleableStringProperty.applyStyle(StyleableStringProperty.java:45)
        at javafx.web/javafx.scene.web.HTMLEditor.<init>(HTMLEditor.java:50)
        at JavaFXMathML.start(JavaFXMathML.java:26)
        at javafx.graphics/com.sun.javafx.application.LauncherImpl.lambda$launchApplication1$9(LauncherImpl.java:846)
        at javafx.graphics/com.sun.javafx.application.PlatformImpl.lambda$runAndWait$12(PlatformImpl.java:451)
        at javafx.graphics/com.sun.javafx.application.PlatformImpl.lambda$runLater$10(PlatformImpl.java:424)
        at java.base/java.security.AccessController.doPrivileged(Native Method)
        at javafx.graphics/com.sun.javafx.application.PlatformImpl.lambda$runLater$11(PlatformImpl.java:423)
        at javafx.graphics/com.sun.glass.ui.InvokeLaterDispatcher$Future.run(InvokeLaterDispatcher.java:96)
        at javafx.graphics/com.sun.glass.ui.gtk.GtkApplication._runLoop(Native Method)
        at javafx.graphics/com.sun.glass.ui.gtk.GtkApplication.lambda$runLoop$11(GtkApplication.java:277)
        ... 1 more

**Comment by kevinrushforth** [comment]
Not sure why you are using make_runargs.sh -- that script is usually meant for taking your own build of JavaFX done on one machine and copying it to another where the path to the FX modules is different (since run.args has absolute paths).

In any case, the problem seems to be that the exports from buildSrc/addExports is not being included in the generated run.args as it should be, which is a bug in make_runargs.sh. You can append the needed exports from that file to your generated run.args, or just not use it in the first place and take the build/run.args from your JavaFX build and manually adjust the paths as needed.

**Comment by scientificware** [comment]
Thanks.
A mistake.
Missing build/run.args from your JavaFX build and manually adjust the paths as needed.

**Comment by scientificware** [comment]
Some working documents :
- [MathML Refactoring in WebKit](http://frederic-wang.fr/mathml-refactoring-in-webkit.html)
- [MathML Improvements in WebKit](http://frederic-wang.fr/mathml-improvements-in-webkit.html)

**Comment by scientificware** [comment]
If you want to build WebKit :

Avant de lancer la construction modifier le fichier  rt/gradle.properties.template :
Enregister le fichier modifié sous : «  rt/gradle.properties.template ».

# These properties give developers the chance to skip building WebKit and/or
# GStreamer. WebKit takes a fair amount of time to build (more than 50% of the
# overall full build time is taken by WebKit), so allowing a developer to
# selectively enable building of WebKit is important. To build WebKit or
# GStreamer, uncomment the appropriate lines below.

COMPILE_WEBKIT = true

**Comment by scientificware** [comment]
Just a problem with a "lstdc++ missing" in my configuration.

To do : modify the required development package list for Mageia 6.

- The build with webkit is now ok.
- Running JavaFX with custom RenderMathMLFraction.cpp where I inflated the fraction line. That doesn't resolve the bug : just an exercice to be familiar with the code.
![screenshot_javafxmathml_20180505_100542](https://user-images.githubusercontent.com/19194678/39661273-1639485c-504f-11e8-88c6-0c051c719467.png)

**Comment by scientificware** [comment]
- The numerator and denominator vertical sizes are equal to 0,
- That's confirms my analysis : the problem is in one super class : May be RenderBox.

**Comment by scientificware** [comment]
https://trac.webkit.org/browser

**Comment by scientificware** [comment]
Change [c++ option](https://gcc.gnu.org/onlinedocs/gcc/C_002b_002b-Dialect-Options.html#C_002b_002b-Dialect-Options)

Where is the best place to change this option ?

-fno-rtti

    Disable generation of information about every class with virtual functions for use by the C++ run-time type identification features (dynamic_cast and typeid). If you don’t use those parts of the language, you can save some space by using this flag. Note that exception handling uses the same information, but G++ generates it as needed. The dynamic_cast operator can still be used for casts that do not require run-time type information, i.e. casts to void * or to unambiguous base classes.

**Comment by scientificware** [comment]
Hi,
Thank you very much,
- I'd already "post-it" these links in previous comment three days ago.
- And I already made the changes you mentioned in WebKitCompilerFlags.cmake.
- I also read changelog. May be I missed some informations.

I want to remove -fno-rtti, in order to use typeid to identify child in RenderMathMLBlock and understand why heights are set to 1.

Have you got a specific configuration (Netbeans or others) to debug JavaFX/WebKit ?

Best regards.

**Comment by scientificware** [comment]
The child of a RenderMathMLBlock is an RenderBlockFlow.

**Comment by scientificware** [comment]
I think that I gone a bit farther. The problem only affects MathML elements.
I'm Trying something rather simple this morning.

**Comment by scientificware** [comment]
I think that MathML Renders are cast in an inappropiate object.

**Comment by scientificware** [comment]
http://doc.qt.io/qt-5/qmake-test-function-reference.html

**Comment by scientificware** [comment]
Do not add these files which are already added in SOURCES
```
        # css/CSSDefaultStyleSheets.cpp \
        # accessibility/AXObjectCache.cpp \
        # accessibility/AccessibilityNodeObject.cpp \
        # accessibility/AccessibilityObject.cpp \
        # accessibility/AccessibilityRenderObject.cpp \
        # dom/Document.cpp \
```

**Comment by scientificware** [comment]
Which part of WebKit use JSMathMLElementWrapperFactory ?

**Comment by scientificware** [comment]
I'm looking for what is the fonction of `WTFMove`.

**Comment by scientificware** [comment]
- Is there an specific implementation of fontCascade witch returns wrong text height value ?
  - See in /modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/java
 (FontCascade, FontCascadeJava).
  - MathML implementation uses fontCascade informations does java implementation returns at least defaults MathML informations ?

- Try to understand this comment in /modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/java/
```
FloatRect Font::platformBoundsForGlyph(Glyph) const
{
    return FloatRect(); //That is OK! platformWidthForGlyph impl is enough.
}
```

**Comment by scientificware** [comment]
```
void RenderMathMLToken::layoutBlock(bool relayoutChildren, LayoutUnit pageLogicalHeight)
{
    ASSERT(needsLayout());

    if (!relayoutChildren && simplifiedLayout())
        return;

    GlyphData mathVariantGlyph;
    if (m_mathVariantCodePoint)
        mathVariantGlyph = style().fontCascade().glyphDataForCharacter(m_mathVariantCodePoint.value(), m_mathVariantIsMirrored);

    if (!mathVariantGlyph.font) {
        RenderMathMLBlock::layoutBlock(relayoutChildren, pageLogicalHeight);
        return;
    }

    for (auto* child = firstChildBox(); child; child = child->nextSiblingBox())
        child->layoutIfNeeded();

    setLogicalWidth(mathVariantGlyph.font->widthForGlyph(mathVariantGlyph.glyph));
    
    setLogicalHeight(mathVariantGlyph.font->boundsForGlyph(mathVariantGlyph.glyph).height());
    clearNeedsLayout();
}
```

**Comment by scientificware** [comment]
To try this weekend :
According my comment above :
For Java platform [platformBoundsForGlyph]( openjdk-jfx/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/java/FontJava.cpp ) (doesn't return the bounds of the glyph but a (0,0,0,0) rectangle.
Then boundsForGlyph returns 0 for the height.

**Comment by scientificware** [comment]
`FontJava.cpp`

```
FloatRect Font::platformBoundsForGlyph(Glyph) const
{
    ~~return FloatRect(); //That is OK! platformWidthForGlyph impl is enough~~.
}
```

**Comment by scientificware** [comment]
First tests confirm my analysis.
Found what's wrong with MathML lines overlapp.
In fact MathML needs `platformBoundsForGlyph` implementation.

`WCFontPerfLogger.java`

```
    public float[] getGlyphBoundingBox(int glyph) {
        logger.resumeCount("GETGLYPHBOUNDINGBOX");
        float[] res = fnt.getGlyphBoundingBox(glyph);
        logger.suspendCount("GETGLYPHBOUNDINGBOX");
        return res;
    }
```

`WCFont.java`

```
public abstract float[] getGlyphBoundingBox(int glyph);
```

**Comment by scientificware** [comment]
Morning correction from previous morning suggestion, before breakfast too : 

[`FontJava.cpp`](https://github.com/javafxports/openjdk-jfx/blob/48fd8216450b58aa100fbacf8d070b0385f0aaf5/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/java/FontJava.cpp)
~~return FloatRect(); //That is OK! platformWidthForGlyph impl is enough~~

```
FloatRect Font::platformBoundsForGlyph(Glyph c) const
    {
    //return FloatRect(); //That is OK! platformWidthForGlyph impl is enough.
    
    JNIEnv* env = WebCore_GetJavaEnv();

    RefPtr<RQRef> jFont = m_platformData.nativeFontData();
    if (!jFont) {
        return FloatRect();
    }

    static jmethodID getGlyphBoundingBox_mID = env->GetMethodID(PG_GetFontClass(env), "getGlyphBoundingBox", "(I)[F");
    ASSERT(getGlyphBoundingBox_mID);

    jfloatArray boundingBox = (jfloatArray)env->CallObjectMethod(*jFont, getGlyphBoundingBox_mID, (jint)c);
    
    jfloat *bBox = env->GetFloatArrayElements(boundingBox,0);
    
    CheckAndClearException(env);

    return FloatRect(bBox[0], bBox[1], bBox[2], bBox[3]);
    }
```

**Comment by scientificware** [comment]
The remaining problem seems to be an ascent and descent Glyph, I'm still working on it.

**Comment by scientificware** [comment]
I'm right !

**Comment by scientificware** [comment]
[`WCFontImpl.java`
](https://github.com/javafxports/openjdk-jfx/blob/48fd8216450b58aa100fbacf8d070b0385f0aaf5/modules/javafx.web/src/main/java/com/sun/javafx/webkit/prism/WCFontImpl.java)

The ascent and descent problem has resolved by :
```
@Override public float[] getGlyphBoundingBox(int glyph) {
        float[] bb = new float[4];
        bb = getFontStrike().getFontResource().getGlyphBoundingBox(glyph, font.getSize(), bb);
        bb[1]= -getAscent();
        bb[3]= getDescent();
        System.out.println("GlyphBoundingBox "+bb[0]+" "+bb[1]+" "+bb[2]+" "+bb[3]);
        return bb;
    }
```

**Comment by kevinrushforth** [comment]
@scientificware if you plan to contribute your fix for this issue to FX, we will need a signed [OCA](http://www.oracle.com/technetwork/community/oca-486395.html) on file before it can be reviewed and accepted; see [CONTRIBUTING.md](https://github.com/javafxports/openjdk-jfx/blob/develop/.github/CONTRIBUTING.md).

**Comment by muralibilla** [comment]
@scientificware   Nice to see that you made good progress on this issue. We can treat this issue separately from webkit upgrade. Once you are ready with complete patch (including unit tests),  you can generate a pull request to initiate review.

**Comment by scientificware** [comment]
Abossolo Foh Guy - NB
I already signed the OCA.

**Comment by kevinrushforth** [comment]
Yes, I see you on the list of OCA signatories. Thanks.

**Comment by scientificware** [comment]
{ 0x28  , 0x239b, 0x239c, 0x239d, 0x0    }, // left parenthesis
    { 0x29  , 0x239e, 0x239f, 0x23a0, 0x0    }, // right parenthesis
    { 0x5b  , 0x23a1, 0x23a2, 0x23a3, 0x0    }, // left square bracket
    { 0x5d  , 0x23a4, 0x23a5, 0x23a6, 0x0    }, // right square bracket
    Name|character|topChar|extensionChar|bottomChar|middleChar
---|---|---|---|---|---
left curly bracket| 0x7b | 0x23a7| 0x23aa| 0x23a9| 0x23a8  
right curly bracket | 0x7d | 0x23ab| 0x23aa| 0x23ad| 0x23ac  
left ceiling | 0x2308| 0x23a1| 0x23a2| 0x23a2| 0x0  
right ceiling | 0x2309| 0x23a4| 0x23a5| 0x23a5| 0x0  
left floor | 0x230a| 0x23a2| 0x23a2| 0x23a3| 0x0  
right floor | 0x230b| 0x23a5| 0x23a5| 0x23a6| 0x0  
vertical bar | 0x7c | 0x7c| 0x7c| 0x7c| 0x0  
double vertical line | 0x2016| 0x2016| 0x2016| 0x2016| 0x0  
parallel to | 0x2225| 0x2225| 0x2225| 0x2225| 0x0  
integral sign | 0x222b| 0x2320| 0x23ae| 0x2321| 0x0

**Comment by scientificware** [comment]
## Yes !
With this small correction matrix delimiters are well displayed !

![screen_javafx_mathml](https://user-images.githubusercontent.com/19194678/41248839-7c3b5d86-6db2-11e8-8ea2-3d955e098046.png)

**Comment by scientificware** [comment]
The remaining problems:
- I'm afraid that without mathtable informations, glyphs assembly for stretchy glyphs couldn't be better and may vary from one font to another.
- Big opérators are too small.
- No horizontal brackets.

![screenshot_20180612_052032](https://user-images.githubusercontent.com/19194678/41268491-9aabf1a6-6e00-11e8-8649-bd5cacad19dc.png)

**Comment by scientificware** [comment]
Ok, I've just understand what's wrong !
JavaFX used the bad code for glyph in latin modern.

Name|Char.|Sy.|Top Char.|Extension Char.|Bottom Char.|Middle Char.
---|---|----|---|---|---|---
left curly bracket| ?| { | ? ⎧ | 0x23aa ⎪ | 0x23a9 ⎩ | 0x23a8 ⎨
right curly bracket | 0x7d| } | 0x23ab ⎫ | 0x23aa ⎪ | 0x23ad ⎭ | 0x23ac ⎬ 
left ceiling | 0x2308| ⌈ | 0x23a1| 0x23a2| 0x23a2| 0x0  
right ceiling | 0x2309| ⌉ | 0x23a4| 0x23a5| 0x23a5| 0x0  
left floor | 0x230a| ⌈ | 0x23a2| 0x23a2| 0x23a3| 0x0  
right floor | 0x230b| ⌋ | 0x23a5| 0x23a5| 0x23a6| 0x0  
vertical bar | 0x7c| \| | 0x7c| 0x7c| 0x7c| 0x0  
double vertical line | 0x2016| ‖ | 0x2016| 0x2016| 0x2016| 0x0  
parallel to | 0x2225| ∥ | 0x2225| 0x2225| 0x2225| 0x0  
integral sign | 0x222b| ∫ | 0x2320| 0x23ae| 0x2321| 0x0

**Comment by scientificware** [comment]
```
/*
 * Copyright (c) 2017, 2018, Oracle and/or its affiliates. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.  Oracle designates this
 * particular file as subject to the "Classpath" exception as provided
 * by Oracle in the LICENSE file that accompanied this code.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 */
package javafxmathml;

import javafx.application.Application;
import javafx.application.Platform;
import javafx.scene.Scene;
import javafx.scene.layout.StackPane;
import javafx.scene.web.HTMLEditor;
import javafx.scene.web.WebEngine;
import javafx.scene.web.WebView;
import javafx.stage.Stage;

/**
 * @test
 * @bug JDK-8147476 Rendering issues with MathML token elements
 * @bug JDK-8089878 HTMLEditor messes up MathML markup press
 */
public class OpenJFXMathMLRenderingIssueTest extends Application {

    final HTMLEditor htmlEditor = new HTMLEditor();
    final StackPane root = new StackPane();
    final WebView webView = (WebView) htmlEditor.lookup(".web-view");

    public void init() {
        htmlEditor.setHtmlText(content);
        root.getChildren().add(htmlEditor);
        htmlEditor.setPrefWidth(400);
        htmlEditor.setPrefHeight(150);
    }

    @Override
    public void start(Stage primaryStage) {

        webView.focusedProperty().addListener((observable, oldValue, newValue) -> {

            if (newValue) {
                webView.getBoundsInParent().toString();

                WebEngine we = webView.getEngine();
                int height = (int) we.executeScript(
                        "elements = document.getElementsByTagName('mo');"
                        + "element = elements[0].clientHeight;"
                );

                if (height < 2) {
                    Platform.exit();
                    throw new RuntimeException(" : MathML rendering issues see JDK-8089878 and JDK-8147476.\n");
                }

                System.out.println("[Ok]\n");
                Platform.exit();
            }
        });

        primaryStage.setTitle("OpenJFX MathML Rendering Issue Test");
        primaryStage.setScene(new Scene(root));
        primaryStage.show();

    }
    /**
     * @param args the command line arguments
     */
    /*public static void main(String[] args) {
        launch(args);
    }*/
    String BodyContent
            = "<math display=\"block\">"
            + "   <mrow>"
            + "      <mi>x</mi>"
            + "      <mo>=</mo>"
            + "      <mfrac>"
            + "         <mrow>"
            + "            <mo>−</mo>"
            + "            <mi>b</mi>"
            + "            <mo>±</mo>"
            + "            <msqrt>"
            + "               <mrow>"
            + "                  <msup>"
            + "                     <mi>b</mi>"
            + "                     <mn>2</mn>"
            + "                  </msup>"
            + "                  <mo>−</mo>"
            + "                  <mn>4</mn>"
            + "                  <mi>a</mi>"
            + "                  <mi>c</mi>"
            + "               </mrow>"
            + "            </msqrt>"
            + "         </mrow>"
            + "         <mrow>"
            + "            <mn>2</mn>"
            + "            <mi>a</mi>"
            + "         </mrow>"
            + "      </mfrac>"
            + "   </mrow>"
            + "</math>";

    String content = "<!doctype html>"
            + "<html>"
            + "   <head>"
            + "      <meta charset=\"UTF-8\">"
            + "      <title>OpenJFX and MathML</title>"
            + "   </head>"
            + "   <body>"
            + "      <p>"
            + BodyContent
            + "      </p>"
            + "   </body>"
            + "</html>";

}
```

**Comment by scientificware** [comment]
Exactly what I need and what I has been waiting since few years !
Hope, I could use it in my projects soon.

![screenshot_20180613_123931](https://user-images.githubusercontent.com/19194678/41346540-0c40b66a-6f07-11e8-814a-2601bf5c77ef.png)

**Comment by scientificware** [comment]
```
/*
 * Copyright (c) 2017, 2018, Oracle and/or its affiliates. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.  Oracle designates this
 * particular file as subject to the "Classpath" exception as provided
 * by Oracle in the LICENSE file that accompanied this code.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 */

import javafx.application.Application;
import javafx.application.Platform;
import javafx.scene.Scene;
import javafx.scene.layout.StackPane;
import javafx.scene.web.HTMLEditor;
import javafx.scene.web.WebEngine;
import javafx.scene.web.WebView;
import javafx.stage.Stage;

/**
 * @test
 * @bug JDK-8147476 Rendering issues with MathML token elements
 * @bug JDK-8089878 HTMLEditor messes up MathML markup press
 */
public class NavigateurTest extends Application {

    final HTMLEditor htmlEditor = new HTMLEditor();
    final StackPane root = new StackPane();
    final WebView webView = (WebView) htmlEditor.lookup(".web-view");

    public void init() {
        htmlEditor.setHtmlText("");
        root.getChildren().add(htmlEditor);
        htmlEditor.setPrefWidth(800);
        htmlEditor.setPrefHeight(600);
    }

    @Override
    public void start(Stage primaryStage) {
    
        webView.getEngine().load("https://duckduckgo.com");

        primaryStage.setTitle("OpenJFX MathML Rendering Issue WebBrowser Test");
        primaryStage.setScene(new Scene(root));
        primaryStage.show();

    }

}
```

**Comment by scientificware** [comment]
[JDK-8147476](https://bugs.openjdk.java.net/browse/JDK-8147476?jql=text%20~%20"Mathml")
Rendering issues with MathML token elements

**Comment by scientificware** [comment]
Another question was not resolved :

Why programmers didn't used the same logic in normal text and mathml text ?
They call bounding box height dimension rather than sum the text ascent and descent.
One who wrapped webkit in JavaFX must have been thinking that it was the same logic.

My response : In fact the problem with mathematical glyphs is that they go upper and lower than ascent and the descent. And only the bounding box height gives this right measure of the glyph.

**Comment by scientificware** [comment]
To do : 
[ ] : Find why cursor disappear when enter in <mo> and become visible in <mn> tokens.

**Comment by scientificware** [comment]
gradle -PFULL_TEST=true :systemTests:test

**Comment by scientificware** [comment]
# Close Buzilla Webkit Report.
![ogimage](https://user-images.githubusercontent.com/19194678/41807124-f594ce36-76c9-11e8-9ea7-9ea63333aad6.png)

https://bugs.webkit.org/show_bug.cgi?id=185417

**Comment by scientificware** [comment]
Bug report telated to this bug :
- [JDK-8090887](https://bugs.openjdk.java.net/browse/JDK-8090887)

**Comment by scientificware** [comment]
Problems with Windows CR when editing files.
Really too hard to work under Windows.
Don't understand the logic of core.autocrlf !
Definitively stays on linux when working on GitHub OpenJFX project.
Not at home install Linux Mageia in a [VirtualBox](https://www.virtualbox.org/).

![vbox_logo2_gradient](https://user-images.githubusercontent.com/19194678/42033674-e411d902-7add-11e8-9ad6-583b3625e416.png)

**Comment by kevinrushforth** [comment]
As long as the end result is the same, you can use any method you like. My suggestion was just an example of one way to do it.

**Comment by scientificware** [comment]
Thank you very much Kevin.
Sure I would never succed without this and would be still stuck with git :-)
That was exactly what I needed.
I'm writing a "How to" and a notes for me. So I post here all helps and suggestions that you and your team gave. And of course all you teached me. With comments for future. I'll hope it will be helpfull for someone else.

**Comment by scientificware** [comment]
gradle :web:test

**Comment by scientificware** [comment]
Something new ?
Unified Source.

**Comment by scientificware** [comment]
Can't build `WebCore` with WebKit 606.1.
And a problem with gradle : can't clean ?
May be because I disable java 9 ?
In fact a problem with `ant`.
```

===================================================================
 WebKit is now built (1h:09m:04s). 
 To run FXLauncher with this newly-built code, use the
 "../../src/main/native/Tools/Scripts/run-minibrowser" script.
====================================================================

> Task :web:compileJavaDOMBindingLinux
Note: Some input files use or override a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
Note: Some input files use unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.

> Task :fxml:compileJava
Note: /home/scientificware2016/Documents/openjdk-jfx-compile/modules/javafx.fxml/src/main/java/javafx/fxml/FXMLLoader.java uses or overrides a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
Note: Some input files use unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.

> Task :swing:compileJava
Note: Some input files use or override a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
Note: Some input files use unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.

BUILD SUCCESSFUL in 1h 13m 38s
124 actionable tasks: 124 executed
```

**Comment by scientificware** [comment]
![screenshot_20180705_202453](https://user-images.githubusercontent.com/19194678/42340982-d68d5c98-8091-11e8-86fa-e961d513d9a2.png)

**Comment by scientificware** [comment]
numerator

**Comment by kevinrushforth** [comment]
This was fixed some time ago by PR #117

**Comment by scientificware** [comment]
@kevinrushforth For a better MathML experience,  a [`mathml.css `](https://github.com/javafxports/openjdk-jfx/blob/fd73ad9a64cd84633968ba757b3f113be0b1a1b3/modules/javafx.web/src/main/native/Source/WebCore/css/mathml.css#L51) comment suggests that at least one of these fonts must be installed (in order of preference). 

- `Latin Modern Math`

- `STIX Two Math`
- `XITS Math`
- `STIX Math`
- `Libertinus Math`
- `TeX Gyre Termes Math`
 
- `TeX Gyre Bonum Math`
- `TeX Gyre Schola`

- `DejaVu Math TeX Gyre`

- `TeX Gyre Pagella Math`
- `Asana Math`

- `Cambria Math`
- `Lucida Bright Math`
- `Minion Math`

- `Times New Roman`

Is it possible to indicate that to users somewhere in the documentation or elsewhere.

Before `Latin Modern Math` install :
![screenshot_20180830_235258](https://user-images.githubusercontent.com/19194678/45088566-a8d2a800-b109-11e8-9bbc-29336bf80d67.png)

After `Latin Modern Math` install : the greek small letter PI is well rendered, better than with the default font.
![screenshot_20180831_013116](https://user-images.githubusercontent.com/19194678/45088614-c99afd80-b109-11e8-9f54-b6d0ecb0704a.png)
